codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -0,0 +1,297 @@
+        if (!streamInfo.inStreamsToFireChildReadComplete) {
+            streamsToFireChildReadComplete.add(streamInfo);
+            streamInfo.inStreamsToFireChildReadComplete = true;
+        }
+    }
+
+    class ConnectionListener extends Http2ConnectionAdapter {
+        @Override
+        public void onStreamActive(Http2Stream stream) {
+            if (ctx == null) {
+                // UPGRADE stream is active before handlerAdded().
+                return;
+            }
+            // If it is an outgoing stream, then we already created the channel.
+            // TODO: support outgoing streams
+            if (stream.getProperty(StreamInfo.class) != null) {
+                return;
+            }
+            stream.setProperty(StreamInfo.class,
+                new StreamInfo(createStreamChannel(ctx, stream, streamHandler)));
+        }
+
+        @Override
+        public void onStreamClosed(Http2Stream stream) {
+            StreamInfo streamInfo = (StreamInfo) stream.removeProperty(StreamInfo.class);
+            if (streamInfo != null) {
+                streamInfo.childChannel.close();
+            }
+        }
+
+        @Override
+        public void onGoAwayReceived(final int lastStreamId, long errorCode, ByteBuf debugData) {
+            final Http2GoAwayMessage goAway = new Http2GoAwayMessage(errorCode, debugData);
+            try {
+                http2Handler.connection().forEachActiveStream(new Http2StreamVisitor() {
+                    @Override
+                    public boolean visit(Http2Stream stream) {
+                        if (stream.id() > lastStreamId
+                                // TODO: enable when syncing to master
+                                /*&& http2Handler.connection().local().isValidStreamId(stream.id())*/) {
+                            StreamInfo streamInfo = (StreamInfo) stream.getProperty(StreamInfo.class);
+                            if (streamInfo != null) {
+                                goAway.retain();
+                                streamInfo.childChannel.pipeline().fireUserEventTriggered(goAway);
+                            }
+                        }
+                        return true;
+                    }
+                });
+            } catch (Http2Exception ex) {
+                ctx.invoker().invokeExceptionCaught(ctx, ex);
+            }
+            // TODO: who releases the original reference to debugData? Does this need a retain?
+            ctx.fireUserEventTriggered(goAway);
+        }
+    }
+
+    class FrameListener extends Http2FrameAdapter {
+        @Override
+        public void onRstStreamRead(ChannelHandlerContext ctx, int streamId, long errorCode)
+                throws Http2Exception {
+            Http2Stream stream = http2Handler.connection().requireStream(streamId);
+            StreamInfo streamInfo = (StreamInfo) stream.getProperty(StreamInfo.class);
+            // Use a user event in order to circumvent read queue.
```

Comment:
Yes, we notify about RST_STREAM immediately because we don't know whether the application will want to throw away what is in the queue or whether it still wants to process it. A server would commonly want to throw it away in order to not waste cycles, but a client may want to process whatever it had received up to that point. Skipping the queue allows the application to decide what to do with the queue.

I'll improve the comment.